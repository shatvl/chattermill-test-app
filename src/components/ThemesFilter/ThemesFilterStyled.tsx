import styled from 'styled-components';

export const FilterWrapper = styled.div`
  position: relative;
  max-width: 40rem;
  margin: auto;
`;

export const FilterInput = styled.input`
  display: block;
  width: 100%;
  height: 3.375rem;
  padding: ${({ theme }) => `calc(.75rem - ${theme.borderWidth}) calc(.9rem - ${theme.borderWidth})`};
  border: ${({ theme }) => theme.borderWidth} solid ${({ theme }) => theme.colors.gray};
  border-radius: 0.25rem;
  color: ${({ theme }) => theme.text};
  outline: none;
  transition: border-color 0.15s ease;
  word-break: break-word;
  font-size: 1rem;
  line-height: 1.25rem;
`;

export const FilterPopupList = styled.ul`
  position: absolute;
  background: ${({ theme }) => theme.colors.white};
  list-style: none;
  width: 20rem;
  min-height: 2.5rem;
  max-height: 15rem;
  overflow-y: auto;
  z-index: ${({ theme }) => theme.zIndex.xxl};
  margin: 0.25rem 0;
  padding: 0.25rem 0;
  border-radius: 0.25rem;
  box-shadow: 0px 5px 30px -15px rgba(0, 0, 0, 0.75);
`;

export const FilterPopupListItem = styled.li`
  font-size: 0.875rem;
  color: ${({ theme }) => theme.colors.textSecondary};
  padding: 0.25rem 0.5rem;
  cursor: pointer;

  &:hover {
    background-color: ${({ theme }) => theme.colors.primaryLight};
    color: ${({ theme }) => theme.colors.text};
  }
`;

export const FilterClear = styled.div`
  position: absolute;
  font-size: 2.25rem;
  font-weight: 300;
  right: 1rem;
  cursor: pointer;
`;

export const FilterHeader = styled.div`
  display: flex;
  align-items: center;
  justify-content: space-between;
`;
