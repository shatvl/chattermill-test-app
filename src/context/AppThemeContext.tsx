import React, { FC, createContext, useContext } from 'react';
import { DefaultTheme, ThemeProvider } from 'styled-components';

interface AppThemeContextProps {}

export interface AppTheme extends DefaultTheme {
  colors: {
    [key: string]: string;
  };
  zIndex: {
    [key: string]: number;
  };
  fonts: {
    header: string;
    main: string;
  };
  borderWidth: string;
}

// global theme for app, TODO: dark / light mode
export const getTheme = (): AppTheme => ({
  zIndex: {
    xs: 1,
    s: 2,
    m: 3,
    lg: 4,
    xlg: 5,
    xxlg: 6,
  },
  colors: {
    background: '#fff',
    primary: '#ff8489',
    primaryLight: 'rgba(255, 132, 136, .2)',
    primaryMedium: 'rgba(255, 132, 136, .5)',
    red: '#d65942',
    gray: '#c4c9cc',
    grayLight: 'e3e3e3',
    text: '#2e3939',
    textLighter: '#2e3939',
    textSecondary: '#707579',
    textMeta: '#bfbfbf',
    black: '#000000',
    white: '#ffffff',
  },
  fonts: {
    header: 'Hind Siliguri',
    main: 'Hind Siliguri',
  },
  borderWidth: '0.0625rem',
});

export const AppThemeContext = createContext<AppThemeContextProps>({});

export const AppThemeProvider: FC = ({ children }) => (
  //we could implement dark / light theme here, or user specific styles base on user session
  <AppThemeContext.Provider value={{}}>
    <ThemeProvider theme={getTheme()}>{children}</ThemeProvider>
  </AppThemeContext.Provider>
);

export const usePortalAppThemeContext = () => useContext(AppThemeContext);
