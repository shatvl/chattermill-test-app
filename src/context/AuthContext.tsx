import React, {
  FC, createContext, useContext, useState, useEffect,
} from 'react';
import jwt_decode from 'jwt-decode';
import { CHATTER_TOKEN_COOKIE, getCoockie } from '../services/cookie';
import { User } from '../models/User.model';

interface AuthContextProps {
  user: User;
  isAuthorized: () => boolean;
}

export const AuthContext = createContext<AuthContextProps>({
  user: null,
  isAuthorized: () => false,
});

export const AuthProvider: FC = ({ children }) => {
  const [user, setUser] = useState<User>();

  // decode token from params or try to retrieve from cookie
  const getUser = (token?: string): User => {
    const tok = token ?? getCoockie(CHATTER_TOKEN_COOKIE);

    if (!tok) {
      return null;
    }

    try {
      const decodedUser: { exp: number; id: string; orig_iat: number } = jwt_decode(tok);

      return decodedUser || null;
    } catch (error) {
      console.error(error);
      return null;
    }
  };

  const isAuthorized = () =>
    // check via getUser() method, because token could be expired and if users won't decoded navigate to login page
    !!getUser();
  useEffect(() => setUser(getUser()), []);

  return <AuthContext.Provider value={{ user, isAuthorized }}>{children}</AuthContext.Provider>;
};

export const useAuthContext = () => useContext(AuthContext);
