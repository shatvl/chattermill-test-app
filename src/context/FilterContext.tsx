import React, {
  FC, createContext, useContext, useState,
} from 'react';

interface FilterContextProps {
  themeId: number;
  setThemeId: (themeId: number) => void;
  clearFilters: () => void;
}

export const FilterContext = createContext<FilterContextProps>({
  themeId: null,
  setThemeId: () => {},
  clearFilters: () => {},
});

export const FilterProvider: FC = ({ children }) => {
  const [themeId, setThemeId] = useState<number>();

  const handleSelectThemeId = (themeId: number) => setThemeId(themeId);

  const handleClearFilters = () => setThemeId(null);

  return (
    <FilterContext.Provider value={{ themeId, setThemeId: handleSelectThemeId, clearFilters: handleClearFilters }}>
      {children}
    </FilterContext.Provider>
  );
};

export const useFilterContext = () => useContext(FilterContext);
