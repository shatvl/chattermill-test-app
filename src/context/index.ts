export { AppThemeContext } from './AppThemeContext';
export { usePortalAppThemeContext } from './AppThemeContext';
