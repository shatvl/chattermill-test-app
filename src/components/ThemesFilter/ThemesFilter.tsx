import React, { FC, useEffect, useState } from 'react';
import { navigate } from 'gatsby';
import { useFilterContext } from '../../context/FilterContext';
import { CHATTER_TOKEN_COOKIE, removeCookie } from '../../services/cookie';
import { useThemesListLazyQuery } from '../../services/theme/useThemesLazyQuery';
import { FilterClear, FilterHeader, FilterInput, FilterPopupList, FilterPopupListItem, FilterWrapper } from './ThemesFilterStyled';
import { Button } from '../Button';
import { ButtonSizes } from '../Button/Button';

// may be this filter should be implemented via any filter control
// which support pagination (infinite scroll), but currently I think it's enough
const PAGE_SIZE = 50;

// parts of this components could be moved to stand alone component (cascader selector)
const ThemesFilter: FC = () => {
  const [focused, setFocused] = useState<boolean>(false);
  const [load, { data, error }] = useThemesListLazyQuery();
  const [value, setValue] = useState<string>('');
  // also we can save it to url query params for sharing page purpose
  const { setThemeId, themeId } = useFilterContext();

  useEffect(() => load({ limit: PAGE_SIZE, offset: 0 }), []);

  useEffect(() => {
    // if theme id is cleared outside this component
    if (!themeId) {
      setValue('');
    }
  }, [themeId]);

  const handleOnInputChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    if (event && event.target) {
      const { value } = event.target;
      setValue(value);

      // auto clear filter if user removes all symbols from input
      if (!value) {
        setThemeId(null);
      }
    }
  };

  const handleFilterItemClick = (themeId: number, themeName: string) => () => {
    setThemeId(themeId);
    setValue(themeName);
  };

  const handleClear = () => {
    setValue('');
    setThemeId(null);
  };

  // blur event fires befor click event on popup list item (TODO: check possibility refactor it to outside click handler)
  const handleOnBlur = () => setTimeout(() => setFocused(false), 200);

  const handleLogout = () => {
    removeCookie(CHATTER_TOKEN_COOKIE);
    navigate('/login');
  };

  if (error) {
    return <div>{error}</div>;
  }

  // filter themes by input value
  const filteredData = value ? data.filter(theme => theme.name.toLowerCase().includes(value.toLowerCase())) : data;

  return (
    <FilterWrapper>
      <FilterHeader>
        <h4>Filter by Theme</h4>
        <Button size={ButtonSizes.XS} onClick={handleLogout}>
          Log Out
        </Button>
      </FilterHeader>
      {(value || themeId) && <FilterClear onClick={handleClear}>&times;</FilterClear>}
      <FilterInput
        value={value}
        placeholder="Type theme name..."
        onChange={handleOnInputChange}
        onFocus={() => setFocused(true)}
        onBlur={handleOnBlur}
      />
      {focused && !!filteredData?.length && (
        <FilterPopupList>
          {filteredData.map(theme => (
            <FilterPopupListItem onClick={handleFilterItemClick(theme.id, theme.name)} key={theme.id}>
              {theme.name}
            </FilterPopupListItem>
          ))}
        </FilterPopupList>
      )}
    </FilterWrapper>
  );
};

export default ThemesFilter;
