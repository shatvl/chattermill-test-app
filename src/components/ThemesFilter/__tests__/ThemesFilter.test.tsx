/* eslint-env jest */
import React from 'react';
import renderer from 'react-test-renderer';
import 'jest-styled-components';

import { wait, waitRenderer } from '../../../testing/wait';
import { AppThemeProvider } from '../../../context/AppThemeContext';
import mockFetch from '../../../testing/mockFetch';
import mockedResponse from '../__mocks__/response.json';
import ThemesFilter from '../ThemesFilter';
import { mount } from 'enzyme';
import { FilterClear, FilterInput, FilterPopupListItem } from '../ThemesFilterStyled';
import { FilterContext } from '../../../context/FilterContext';

describe('renders themes filter', () => {
  it('themes filter snapshot', async () => {
    const tree = renderer.create(
      <AppThemeProvider>
        <ThemesFilter />
      </AppThemeProvider>,
    );

    mockFetch(200, mockedResponse);

    await waitRenderer();

    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('themes filter focus input and check items', async () => {
    mockFetch(200, mockedResponse);

    const component = mount(
      <AppThemeProvider>
        <ThemesFilter />
      </AppThemeProvider>,
    );

    await wait();

    component.update();
    expect(component.find(FilterPopupListItem).length).toEqual(0);
    component.find(FilterInput).simulate('focus');
    component.update();
    expect(component.find(FilterPopupListItem).length).toEqual(mockedResponse.data.length);
    component.find(FilterInput).simulate('blur');
  });

  it('themes filter check select / clear', async () => {
    mockFetch(200, mockedResponse);

    const clearFn = jest.fn();
    const setThemeIdFn = jest.fn();

    const component = mount(
      <AppThemeProvider>
        <FilterContext.Provider value={{ clearFilters: clearFn, setThemeId: setThemeIdFn, themeId: null }}>
          <ThemesFilter />
        </FilterContext.Provider>
      </AppThemeProvider>,
    );

    await wait();

    // check select
    component.update();
    component.find(FilterInput).simulate('focus');
    component.update();
    component.find(FilterPopupListItem).first().simulate('click');
    expect(setThemeIdFn).toBeCalledWith(mockedResponse.data[0].id);
    component.update();
    expect(component.find(FilterInput).prop('value')).toEqual(mockedResponse.data[0].name);

    // check clear
    component.find(FilterClear).simulate('click');
    component.update();
    expect(setThemeIdFn).toBeCalledWith(null);
    expect(component.find(FilterInput).prop('value')).toEqual('');
  });
});
