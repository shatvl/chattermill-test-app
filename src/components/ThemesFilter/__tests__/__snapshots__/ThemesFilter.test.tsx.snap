// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders themes filter themes filter snapshot 1`] = `
.c0 {
  position: relative;
  max-width: 40rem;
  margin: auto;
}

.c3 {
  display: block;
  width: 100%;
  height: 3.375rem;
  padding: calc(.75rem - 0.0625rem) calc(.9rem - 0.0625rem);
  border: 0.0625rem solid #c4c9cc;
  border-radius: 0.25rem;
  outline: none;
  -webkit-transition: border-color 0.15s ease;
  transition: border-color 0.15s ease;
  word-break: break-word;
  font-size: 1rem;
  line-height: 1.25rem;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c2 {
  display: inline-block;
  width: initial;
  vertical-align: middle;
  background: transparent;
  border: 1px solid transparent;
  font-size: 0.875rem;
  line-height: 1rem;
  padding: 0.25rem 0.375rem;
  border-radius: 0.25rem;
  color: #ff8489;
  box-sizing: border-box;
  outline: none;
  cursor: pointer;
  -webkit-transition: 0.15s all ease-in-out;
  transition: 0.15s all ease-in-out;
  -webkit-text-decoration: none;
  text-decoration: none;
  white-space: nowrap;
}

.c2:hover,
.c2:active {
  outline: none;
}

.c2:active {
  background-color: rgba(255,132,136,.5);
  border-color: rgba(255,132,136,.2);
}

.c2:disabled,
.c2:disabled:hover,
.c2:disabled:focus,
.c2:disabled:active {
  cursor: not-allowed;
  color: #c4c9cc;
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    <h4>
      Filter by Theme
    </h4>
    <button
      className="c2"
      onClick={[Function]}
      size="xs"
    >
      Log Out
    </button>
  </div>
  <input
    className="c3"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    placeholder="Type theme name..."
    value=""
  />
</div>
`;
